lib/util.c

```diff
@@ -2187,9 +2194,324 @@ EXPORTED void xsyslog_fn(int priority, const char *description,
+{
+    va_list ap;
+    va_start(ap, event_name);
+    char *s = xsyslog_ev_fmt_va(event_name, ap);
+    va_end(ap);
+    return s;
+}
+
+/* xsyslog_ev_fmt_va: va_list version of xsyslog_ev_fmt
+ *
+ * same as xsyslog_ev_fmt, except it gets an initialized va_list argument
+ * instead of a variable argument list
+ *
+ * The CALLER is responsible for freeing the returned string.
+ */
+static char *xsyslog_ev_fmt_va(const char *event_name, va_list ap)
+{
+    struct buf log = BUF_INITIALIZER;
+    struct buf val = BUF_INITIALIZER;
+    const char *sep = " ";  // string that separates kev=value pairs
+    const char *key;
+
+    // Add "event=(event-name)" to the log buffer
+    buf_setcstr(&val, event_name);
+    buf_printf(&log, "event=%s", xsyslog_ev_escape_value(&val));
+
+    /*
+     * Now handle the remaining arguments, which should come in threes:
+     *   key, format, value
+     *
+     * value will be buf_printf'ed with the format,
+     * and then the result will be escaped (if necessary)
+     *
+     * finally, the resulting string "key=escaped-formatted-value"
+     * will be appended to the log buffer
+     */
+    while ((key = va_arg(ap, const char *))) {
+        // special case: __sep__ means reconfigure the separator
+        if (!strcmp(key, "__sep__")) {
+            sep = va_arg(ap, const char *);
+        }
+        else {
+            const char *fmt = va_arg(ap, const char *);
+            va_list val_ap;
+
+            va_copy(val_ap, ap);
+            buf_reset(&val);
+            buf_vprintf(&val, fmt, val_ap);
+            va_end(val_ap);
+
+            buf_printf(&log, "%s%s=%s",
+                       sep, key, xsyslog_ev_escape_value(&val));
+
+            /* Unfortunately neither ap nor val_ap is now pointing to the next 'key' argument.
+             * .. val_ap has been passed to vsprintf, which renders is unfit for further use.
+             .. ap is still pointing to 'val'.
+
+             This function understands `fmt` well enough that it can discard the
+             correct `val` argument and advance `ap` to point to the next `key`.
+            */
+            xsyslog_ev_fmt_skip_va(fmt, ap);
+
+        }
+    }
+
+    buf_free(&val);
+
+    return buf_release(&log);
+}
+
+static void xsyslog_ev_fmt_skip_va(const char* fmt, va_list ap)
+{
+    switch (xsyslog_ev_guess_printf_escape(fmt)) {
+    case 'd': (void) va_arg(ap, int);      break;
+    case 'l': (void) va_arg(ap, long);     break;
+    case 'f': (void) va_arg(ap, double);   break;
+    case 's': (void) va_arg(ap, char *); break;
+    case 'p': (void) va_arg(ap, void *); break;
+    case 0:
+        // we could handle this by skipping the va_arg call entirely, which would allow the user to write
+        //     xsyslog_ev(..., key, "foo", ...)
+        // in place of
+        //     xsyslog_ev(..., key, "foo", DUMMY, ...)
+        // but that seems morelikely to be confusing than otherwise
+    case -1:
```

Comment:
I can do this.  Would it also work to declare the function as returning `signed char`?

Reply:
Also, it's not clear to me that the -1 return is ever useful, and I considered removing it, since the current code doesn't distinguish it from 0. But my heart told me that it was reasonable to distinguish "unparseable" from "found no escape sequences" even if no current code cared about it.